### PR TITLE
[Android Material] Linear Progress Indicator

### LINKED ISSUE #5008: [Material] Linear Progress Indicator
# Material Linear Progress Indicator

https://material.io/develop/ios/components/progress-indicators/progress-views/

- **Visually Identical**
   The rendering should be such that the naked eye can’t tell the difference at a glance. _Not to be confused with “pixel perfect”_

ProgressBar doesn't expose very many apis so the expectation here is that the default would size and color itself the same on iOS and Android

The hopefully not too tricky part here will be that there is only a Material ProgressBar iOS. On the android one it's still just using the original control so material visual on Android will just apply default styles or just inherit from the existing renderer

#### Current state
image

# Scenarios

```XAML
<ProgressBar Visual="Material" />
```

# Difficulty : medium

### PR DESCRIPTION
**Pull request by paymicro**
### Description of Change ###

fixes Linear Progress Indicator

### Issues Resolved ### 

- fixes #5008

### API Changes ###
 
 None

### Platforms Affected ### 

- Android

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

- run [Material] ProgressBar Gallery
- check Visual Gallery

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by PureWeen** [comment]
@paymicro can you rebase this to master when you get a chance?

**Comment by PureWeen** [comment]
build --uitests

File: Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs [hunk 1]
```diff
@@ -13,13 +14,14 @@
...
 
-[assembly: ExportRenderer(typeof(Xamarin.Forms.ProgressBar), typeof(MaterialProgressBarRenderer), new[] { typeof(VisualRendererMarker.Material) })]
+[assembly: ExportRenderer(typeof(ProgressBar), typeof(MaterialProgressBarRenderer), new[] { typeof(VisualRendererMarker.Material) })]
 
 namespace Xamarin.Forms.Platform.Android.Material
 {
 	public class MaterialProgressBarRenderer : AProgressBar,
 		IVisualElementRenderer, IViewRenderer, ITabStop
 	{
+		const float BackgroundAlpha = 0.3f;
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs]: Where did we get this value from?

File: Xamarin.Forms.Platform.Android/Resources/values/styles.xml [hunk 2]
```diff
@@ -2,7 +2,10 @@
 <resources>
   <style name="XamarinFormsMaterialTheme" parent="Theme.MaterialComponents.Light.DarkActionBar">
   </style>
-  <style name="XamarinFormsMaterialProgressBarHorizontal" parent="Widget.AppCompat.ProgressBar.Horizontal">
+  <style name="XamarinFormsMaterialProgressBarHorizontal" parent="android:Widget.ProgressBar.Horizontal">
+    <item name="android:indeterminateOnly">false</item>
+    <item name="android:minHeight">4dp</item>
+    <item name="android:progressDrawable">@drawable/MaterialProgressBar</item>
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.Android/Resources/values/styles.xml]: Why did you opt to use the base control and a custom drawing?

File: Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs [hunk 3]
```diff
@@ -142,51 +146,11 @@ void UpdateColors()
...
-				ProgressBackgroundTintList = ColorStateList.ValueOf(background);
-
-				// TODO: Potentially override alpha to match material design.
-				ProgressBackgroundTintMode = PorterDuff.Mode.SrcOver;
-			}
+			var primary = Element.ProgressColor.IsDefault ? _defaultPrimaryColor : Element.ProgressColor.ToAndroid();
+			ProgressTintList = ColorStateList.ValueOf(primary);
+			var background = Element.BackgroundColor.IsDefault ? _defaultBackgroundColor : Element.BackgroundColor.ToAndroid();
+			ProgressBackgroundTintList = ColorStateList.ValueOf(background);
+			ProgressBackgroundTintMode = PorterDuff.Mode.SrcIn;
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs]: I think we need to have a fallback to follow material and if only the progress color is set, then we have the lighter background of that color - unless specifically set. Right now, it looks that if I set the color to purple, only the main color is set, the track remains black - immediately breaking the material design. If we want to allow the user to break the guides, then we need to let them do the work to get there.

File: Xamarin.Forms.Platform.Android/Resources/values/styles.xml (hunk 2 above)
Comment (mattleibow) [reply Xamarin.Forms.Platform.Android/Resources/values/styles.xml]: Just got reminded by Shane, Google has a _HARD_ set for 4dp. Of course, this is absolutely not true for iOS, but hey, why follow your own guides.

**Comment by mattleibow** [comment]
@paymicro Could you attach screenshots of before/after to the issue for comparison and records (for when Google breaks the controls and the style suddenly changes under us)?

File: Xamarin.Forms.Platform.Android/Resources/drawable/MaterialProgressBar.xml [hunk 4]
```diff
@@ -0,0 +1,51 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!-- 
+From original drawable https://android.googlesource.com/platform/frameworks/base/+/master/core/res/res/drawable/progress_horizontal.xml 
+with removed rounded corners.
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.Android/Resources/drawable/MaterialProgressBar.xml]: Maybe use the AppCompat - if possible - and use this style: https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/res/res/drawable/progress_horizontal_material.xml

File: Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs (hunk 1 above)
Comment (mattleibow) [reply Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs]: Yay. Google. This is typical for Material - neither platform matches the spec.

File: Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs [hunk 5]
```diff
@@ -13,13 +13,14 @@
...
 
-[assembly: ExportRenderer(typeof(Xamarin.Forms.ProgressBar), typeof(MaterialProgressBarRenderer), new[] { typeof(VisualRendererMarker.Material) })]
+[assembly: ExportRenderer(typeof(ProgressBar), typeof(MaterialProgressBarRenderer), new[] { typeof(VisualRendererMarker.Material) })]
 
 namespace Xamarin.Forms.Platform.Android.Material
 {
 	public class MaterialProgressBarRenderer : AProgressBar,
 		IVisualElementRenderer, IViewRenderer, ITabStop
 	{
+		const float BackgroundAlpha = 0.6f;
```
Comment (mattleibow) [inline Xamarin.Forms.Platform.Android/Material/MaterialProgressBarRenderer.cs]: I think this constant should be moved to the `MaterialColors` type (as the rest of the values like this are there)

**Comment by mattleibow** [comment]
Merged this into the material branch for the slider / progress as they are the same control and I want to reuse some of the coloring code. (and the nice gallery)

**Pull request merged**